src/components/application_manager/src/rpc_protection_manager_impl.cc

```diff
@@ -0,0 +1,218 @@
+
+  return encryption_needed_cache_.find(std::make_pair(
+             app_id, correlation_id)) != encryption_needed_cache_.end();
+}
+
+bool RPCProtectionManagerImpl::IsExceptionRPC(
+    const uint32_t function_id) const {
+  using namespace rpc_encryption_exceptions;
+  const std::string policy_fucntion_id = policy_table::EnumToJsonString(
+      static_cast<policy_table::FunctionID>(function_id));
+  return helpers::in_range(kExceptionRPCs, policy_fucntion_id);
+}
+
+void RPCProtectionManagerImpl::AddToEncryptionNeededCache(
+    const uint32_t app_id, const uint32_t correlation_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock lock(message_needed_encryption_lock_);
+
+  LOG4CXX_DEBUG(logger_, "Adding rpc with correlation id: " << correlation_id);
+
+  encryption_needed_cache_.insert(std::make_pair(app_id, correlation_id));
+}
+
+void RPCProtectionManagerImpl::RemoveFromEncryptionNeededCache(
+    const uint32_t app_id, const uint32_t correlation_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock lock(message_needed_encryption_lock_);
+
+  LOG4CXX_DEBUG(logger_,
+                "Removing rpc with correlation id: " << correlation_id);
+
+  encryption_needed_cache_.erase(std::make_pair(app_id, correlation_id));
+}
+
+void RPCProtectionManagerImpl::OnPTUFinished(const bool ptu_result) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  sync_primitives::AutoLock lock(encrypted_rpcs_lock_);
+
+  if (ptu_result) {
+    LOG4CXX_TRACE(logger_,
+                  "PTU finished successfully, commencing internal encrypted "
+                  "RPC cache update");
+    encrypted_rpcs_.clear();
+    SaveEncryptedRPC();
+  } else {
+    LOG4CXX_WARN(logger_,
+                 "PTU was unsuccessful. Keeping internal RPC cache from "
+                 "current snapshot");
+  }
+}
+
+void RPCProtectionManagerImpl::SaveEncryptedRPC() {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  const auto policy_encryption_flag_getter =
+      policy_handler_.PolicyEncryptionFlagGetter();
+
+  const auto policy_policy_app_ids =
+      policy_encryption_flag_getter->GetApplicationPolicyIDs();
+
+  for (const auto& app : policy_policy_app_ids) {
+    LOG4CXX_DEBUG(logger_, "Processing app name: " << app);
+
+    encrypted_rpcs_[app] = GetEncryptedRPCsForApp(app);
+  }
+}
+
+void RPCProtectionManagerImpl::OnPTInited() {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  encrypted_rpcs_.clear();
+
+  SaveEncryptedRPC();
+}
+
+RPCProtectionManagerImpl::FunctionNames
+RPCProtectionManagerImpl::GetEncryptedRPCsForApp(
+    const std::string& policy_app_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  FunctionNames encrypted_rpcs;
+
+  const auto policy_encryption_flag_getter =
+      policy_handler_.PolicyEncryptionFlagGetter();
+
+  if (!policy_encryption_flag_getter->AppNeedEncryption(policy_app_id)) {
+    return encrypted_rpcs;
+  }
+
+  const auto function_groups =
+      policy_encryption_flag_getter->GetFunctionGroupsForApp(policy_app_id);
```

Comment:
Small typo sorry,  I believe it should be `GetFunctionalGroupsForApp`

Reply:
Also wondering if the returned functional groups will also contain the default policies in addition to the app specified ones?

Reply:
Renamed in 988d4fca0c6a9825e3e2c301905c00705bbcc901 . Yes, this will contain default policies, to which sdl falls back if there are no specific app policies